src/libraries/Microsoft.PowerFx.Interpreter/EvalVisitor.cs

```diff
@@ -300,40 +394,59 @@ public override async ValueTask<FormulaValue> Visit(CallNode node, EvalVisitorCo
                     args[i] = await child.Accept(this, context.IncrementStackDepthCounter()).ConfigureAwait(false);
                 }
                 else
-                {
+                {
+                    // This is where Lambdas are created. They close over key values to invoke.
                     args[i] = new LambdaFormulaValue(node.IRContext, child, this, context);
                 }
             }
 
             var childContext = context.SymbolContext.WithScope(node.Scope);
 
             FormulaValue result;
+
+            // Remove this: https://github.com/microsoft/Power-Fx/issues/2821
             IReadOnlyDictionary<TexlFunction, IAsyncTexlFunction> extraFunctions = _services.GetService<IReadOnlyDictionary<TexlFunction, IAsyncTexlFunction>>();
 
             try
             {
-                if (func is IAsyncTexlFunction asyncFunc || extraFunctions?.TryGetValue(func, out asyncFunc) == true)
+                IFunctionInvoker invoker = GetInvoker(func);
+
+                // Standard invoke path. Make everything go through here. 
+                // Eventually collapse all cases to this. 
+                if (invoker != null)
                 {
-                    result = await asyncFunc.InvokeAsync(args, _cancellationToken).ConfigureAwait(false);
+                    var invokeInfo = new FunctionInvokeInfo
+                    {
+                        Args = args,
+                        FunctionServices = _services,
+                        Runner = this,
+                        Context = context.IncrementStackDepthCounter(childContext),

```

Comment:
>IncrementStackDepthCounter [](http://example.com/codeflow?start=42&length=26)

wondering how we got away not increasing the stack depth here. i see we increase it when we visit the arg above which makes sense ,

this is just one step in -> invoke and out right ?

do we have test around max stack depth ?

Reply:
Calling IncrementStackDepthCounter will test max depth and throw at the limit. 

This is another benefit of unifying the invoke paths - we ensure consistent handling here across all paths.